File: src/backend/models/baking.py
"""SQLAlchemy models for Baking Studio module."""
from sqlalchemy import (
    Boolean,
    Column,
    DateTime,
    Float,
    ForeignKey,
    Integer,
    String,
    Text,
    func,
)
from sqlalchemy.orm import relationship
from sqlalchemy.types import JSON

from src.backend.database import Base


class BakingRecipe(Base):
    """A baking recipe with metadata and scaling info."""

    __tablename__ = "baking_recipes"

    id = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String, nullable=False)
    name_zh = Column(String, nullable=True)
    cake_type = Column(String, nullable=False)  # basque | cheesecake | chiffon | cream_cake
    category = Column(String, nullable=False)  # base | cream | decoration | complete
    size = Column(String, default="6inch")  # 4inch | 6inch | universal
    format = Column(String, default="full")  # full | box
    steps = Column(JSON, nullable=True)
    notes = Column(Text, nullable=True)
    is_preset = Column(Boolean, default=False)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())

    ingredients = relationship(
        "BakingIngredient",
        backref="recipe",
        cascade="all, delete-orphan",
        order_by="BakingIngredient.sort_order",
    )


class BakingIngredient(Base):
    """An ingredient line within a baking recipe."""

    __tablename__ = "baking_ingredients"

    id = Column(Integer, primary_key=True, autoincrement=True)
    recipe_id = Column(
        Integer,
        ForeignKey("baking_recipes.id", ondelete="CASCADE"),
        nullable=False,
    )
    name = Column(String, nullable=False)
    name_zh = Column(String, nullable=True)
    amount = Column(Float, nullable=False)
    unit = Column(String, nullable=False)  # g | ml | pc | tsp | tbsp
    group_name = Column(String, default="main")  # main | topping | frosting | filling
    sort_order = Column(Integer, default=0)
    is_scalable = Column(Boolean, default=True)


class HomeInventory(Base):
    """Tracks baking supplies available at home."""

    __tablename__ = "baking_home_inventory"

    id = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String, nullable=False, unique=True)
    name_zh = Column(String, nullable=True)
    category = Column(String, default="decoration")  # decoration | staple | dairy | fruit
    in_stock = Column(Boolean, default=True)
    amount = Column(Float, nullable=True)
    unit = Column(String, nullable=True)
    notes = Column(Text, nullable=True)

File: src/frontend/src/types/baking.ts
export type CakeType = "basque" | "cheesecake" | "chiffon" | "cream_cake";
export type CakeCategory = "base" | "cream" | "decoration" | "complete";
export type CakeSize = "4inch" | "6inch" | "8inch" | "universal";
export type CakeFormat = "full" | "box";
export type IngredientUnit = "g" | "ml" | "pc" | "tsp" | "tbsp";

export interface BakingIngredient {
  id: number;
  recipe_id: number;
  name: string;
  name_zh: string | null;
  amount: number;
  unit: string;
  group_name: string;
  sort_order: number;
  is_scalable: boolean;
}

export interface BakingRecipe {
  id: number;
  name: string;
  name_zh: string | null;
  cake_type: CakeType;
  category: CakeCategory;
  size: CakeSize;
  format: CakeFormat;
  steps: string[] | null;
  notes: string | null;
  is_preset: boolean;
  ingredients: BakingIngredient[];
  created_at: string;
  updated_at: string;
}

export interface RecipeCreatePayload {
  name: string;
  name_zh?: string;
  cake_type: CakeType;
  category: CakeCategory;
  size?: CakeSize;
  format?: CakeFormat;
  steps?: string[];
  notes?: string;
  ingredients: Omit<BakingIngredient, "id" | "recipe_id">[];
}

export interface ScaleRequest {
  anchor_ingredient_id: number;
  target_amount: number;
}

export interface ScaleResponse {
  recipe_id: number;
  scale_factor: number;
  ingredients: BakingIngredient[];
}

export interface HomeInventoryItem {
  id: number;
  name: string;
  name_zh: string | null;
  category: string;
  in_stock: boolean;
  amount: number | null;
  unit: string | null;
  notes: string | null;
}

export interface RecipeFilters {
  cake_type?: CakeType;
  category?: CakeCategory;
  size?: CakeSize;
  format?: CakeFormat;
}
